### ISSUE TITLE
[Perf][B300] Evaluate tiled MoBA routing and direct route packing

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/flash-linear-attention#7]
## Status

This issue records an unmeasured B300 optimization hypothesis. The expected mechanism and thresholds below are acceptance criteria, not observed performance claims.

Baseline: `upstream/main@e47d5d20aeb5989b58a3738b872e7c288a9fb75f`.

## Problem

`parallel_moba` is the dependency-light default path used by the public MoBA layer (`moba_chunk_size=256`, `moba_topk=4`, `use_flash_moba=False`). Its routing preparation currently materializes a full fp32 score tensor over `(target_chunk, head, token)`, two similarly shaped boolean masks, filtered and repeated KV buffers, a `nonzero` route list, and several generic gather/scatter/reduction operations around the two FlashAttention calls.

The score and mask tensors grow with both token count and target-chunk count. At `T=64K`, `H=32`, and chunk size 256, the score tensor alone is roughly 2 GiB, before the masks and packing buffers. The hypothesis is that this routing/preparation/merge pipeline is a material latency and peak-memory bottleneck on B300 at 32K-64K context lengths.

## Bounded scope

First create a fixture-only commit containing an independent small-shape oracle and a repository-native `parallel_moba` benchmark registration. Production code must remain byte-for-byte identical to upstream in this commit; record it as `BENCH_BASE` and freeze the oracle before any kernel change.

Profile `BENCH_BASE` before implementation. Stop and record a no-go if routing preparation plus output merge accounts for less than 18% of full forward latency or less than 12% of full forward+backward latency at either the 32K or 64K default endpoint, or if peak-memory measurements do not show the expected score/mask allocation.

If the viability gate passes, implement one coherent optimized preparation/merge path:

1. Compute target-block mean keys directly from the original K tensor, without constructing `filtered_kv` for routing.
2. Score query and candidate-block tiles in Triton, apply packed-sequence and causal masks in-kernel, and produce the exact global `topk-1` selected earlier blocks. Keep top-k state on-chip where practical or merge explicitly bounded partial candidates. Do not materialize the full score or boolean mask tensors. If partial-candidate scratch is used, report its exact scaling and peak size; do not claim that quadratic scratch is eliminated unless the implementation actually proves that bound.
3. Build route counts, compact expert metadata, packed queries, original `(token, head, route)` positions, and packed expert KV directly from the selected block IDs. Route ordering/position assignment must be stable enough to preserve repeatable forward and backward results; an atomic position scheme without a repeatability proof is not sufficient.
4. Replace the generic `index_reduce`/`index_add` output chain with a fixed-fan-in online-softmax merge over the chunk-local result and at most `topk-1` selected-block results.
5. Make backward data movement explicit. The current PyTorch gather/reshape graph scatters `dmq`, `dmk`, and `dmv` back to original Q/K/V. Any raw Triton packing path must preserve that behavior with deterministic/repeatable unpack-and-add logic while retaining both existing FlashAttention backward calls and `deterministic=True`.
6. Use int64 base, row, token, chunk, and stride arithmetic throughout large-address paths.

Retain the current PyTorch preparation/merge path as the correctness and portability fallback for unsupported or measured-losing shapes. Optimize only the existing MHA contract and measured D/chunk/top-k ranges. Do not add GQA support, replace either FlashAttention kernel, change attention math or precision, change the public API/model/config, touch decode, or modify the optional `use_flash_moba=True` backend.

Expected production files are `fla/ops/moba/parallel.py` and a focused new `fla/ops/moba/routing.py`. Fixture files may include `tests/ops/test_moba.py`, `benchmarks/ops/registry.py`, and a small benchmark adapter if the registry cannot express packed inputs cleanly.

## Frozen correctness contract

Before optimization, establish that unmodified production code passes the complete updated fixture, then freeze the fixture commit, reference, seeds, dtypes, numeric flags, assertions, and tolerances.

The independent PyTorch oracle must build block means and selected blocks using the current MoBA definition, construct each token/head's union of local causal keys and selected earlier blocks, and compute dense masked softmax/value accumulation without calling optimized routing. Use seeded continuous inputs to avoid top-k ties.

The targeted fixture must compare output and `dq/dk/dv` for actual sparse `topk=2/4`, fp16 and bf16, D=64/128, chunk size 64/128/256, dense and unequal packed sequences with non-multiple tails, single-chunk/no-target input, and `topk` larger than the available causal-block count. Keep the existing 0.005 comparison tolerance, existing full-attention-equivalence tests, topk=1 short circuit, validation cases, and NaN-poisoning behavior unchanged.

Promotion gates:

```bash
pytest tests/ops/test_moba.py
pytest tests/models/test_modeling_moba.py
```

Also run repeated identical-input forward/backward checks on the optimized route packing to detect atomic-order drift. No reference, tolerance, seed, dtype, numeric-flag, or environment relaxation is allowed after `BENCH_BASE`.

## B300 measurement plan

Register the public endpoint for bf16 default shapes at T=8K/16K/32K/64K, H=32, D=64/128, chunk size 256, topk 4, plus unequal packed 32K/64K totals. Benchmark both forward and forward+backward against `BENCH_BASE` with identical inputs, warmup, repetitions, clocks, and fully warmed autotune caches in the same B300 allocation.

Use the repository runner for authoritative endpoint comparisons and report sample count, median, mean, std, min, p10, p90, equal-weight geomean, peak allocated/reserved memory, and kernel count. Keep raw JSON/logs and exact commands, environment versions, commit IDs, and Slurm job IDs outside git.

Use torch.profiler/NVTX to separate block-mean, score/top-k, packing, both FlashAttention calls, and output merge. Use full and source-correlated NCU captures on representative 32K/64K dense and unequal-packed shapes. Compare summed duration, launch count, DRAM/L2 traffic, tensor/SM throughput, achieved occupancy, registers/spills, eligible warps, and dominant stalls. NCU must confirm that removed score/mask traffic and packing/merge work explain any endpoint gain.

## Acceptance criteria

- Complete frozen op and model gates pass, with exact shape/dtype/API behavior and finite in-tolerance Q/K/V gradients.
- Full forward geomean is at least 1.10x and full forward+backward geomean at least 1.06x across default 32K/64K shapes; each 32K and 64K endpoint is at least 1.05x in both modes.
- Retained 8K/16K and fallback shapes do not regress by more than 3% outside measured noise.
- The optimized path removes the full `(target_chunk, head, token)` score/mask allocations, reduces routing/preparation peak temporary bytes by at least 4x, and reduces full-op peak allocated memory by at least 20% at T=64K.
- No hidden device-host synchronization, skipped routes, unstable packing, changed attention math, or unexplained clock/measurement artifact accounts for the result.

If the profile-first gate, frozen oracle, endpoint performance thresholds, or memory thresholds fail, record the no-go without widening into custom attention kernels, GQA, decode, model changes, or the external FlashMoBA backend.

## Upstream overlap

No open upstream issue, PR, or remote branch currently targets MoBA routing/packing/merge performance. Merged MoBA work #840/#845 introduced the current operator and model; closed #956 is documentation only. Merged NSA work #961 added a reusable CSR inversion utility, but it does not optimize MoBA and its arbitrary within-block order is not by itself a proof of stable MoBA route packing. Open benchmark-profile PR #1144 may create a mechanical registry conflict only.

**Comment by heiheiha798** [comment]
## Cycle 6 result: environment/dependency abort

This evaluation stopped before production implementation because the required FlashAttention dependency could not be established in the available B300 environment. This is an environment/dependency incapability, not a MoBA kernel correctness or performance no-go.

- Source baseline: `e47d5d20aeb5989b58a3738b872e7c288a9fb75f`.
- Fixture-only `BENCH_BASE`: `3b291467932263f4f733352e8cbebae4261c365f`. This commit adds the independent sparse oracle and benchmark registration while leaving production MoBA source byte-for-byte identical to the source baseline.
- Slurm job 5159 was an invalid pre-gate harness attempt: an abbreviated/full commit-ID comparison failed before any test or GPU operator work.
- Slurm job 5161 confirmed the environment but skipped all 16 operator cases and all 7 model cases because `flash_attn` was absent. Those skips are not treated as a correctness gate.
- Slurm job 5163 performed the single approved isolated dependency attempt. No compatible prebuilt wheel was available for Python 3.12, CUDA 13, and PyTorch 2.13, so `flash-attn==2.8.3.post1` built from source. The build failed with `No space left on device` while writing an nvcc fatbin in the job-local temporary filesystem.

The FlashAttention import smoke, complete operator/model gates, baseline profile, NCU collection, viability classification, and candidate implementation therefore never ran. No production source was changed, no branch was pushed, and no draft PR was created. These results support no technical conclusion about the proposed tiled routing/direct packing design or its potential B300 performance.

Raw logs and the frozen task contract remain in the local ignored `profile/parallel_moba-opt/` evidence workspace.
